/warnaserror-:RuleId should not turn on disabled by default rules to warnings

**Version Used**: Latest VS2019 16.8 P2 dogfood build

**Steps to Reproduce**:

1. Create a C# class library with below code:
```csharp
class C
{
    void M(int x)
    {
    }
}
```
2. Add a NuGet package reference to `Microsoft.CodeAnalysis.NetAnalyzers` package: https://dev.azure.com/dnceng/public/_packaging?_a=package&feed=dotnet5&package=Microsoft.CodeAnalysis.NetAnalyzers&protocolType=NuGet&version=5.0.0-rc2.20420.6
3. Build project and confirm no warnings/errors
4. Add the following property to MSBuild project file, which adds `/warnaserror-:CA1801` to the command line:
```xml
<WarningsNotAsErrors>CA1801</WarningsNotAsErrors>
```

**Expected Behavior**:
No CA1801 warnings/errors

**Actual Behavior**:
```
warning CA1801: Parameter x of method M is never used. Remove the parameter or use it in the method body.
```
Rule `CA1801` is disabled in the analyzer package, `/warnaserror-` should not enable it

--- a/src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs
+++ b/src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs
@@ -12290,6 +12290,44 @@ generated_code = auto");
             }
         }
 
+        [WorkItem(47017, "https://github.com/dotnet/roslyn/issues/47017")]
+        [CombinatorialData, Theory]
+        public void TestWarnAsErrorMinusDoesNotEnableDisabledByDefaultAnalyzers(DiagnosticSeverity defaultSeverity, bool isEnabledByDefault)
+        {
+            // This test verifies that '/warnaserror-:DiagnosticId' does not affect if analyzers are executed or skipped..
+            // Setup the analyzer to always throw an exception on analyzer callbacks for cases where we expect analyzer execution to be skipped:
+            //   1. Disabled by default analyzer, i.e. 'isEnabledByDefault == false'.
+            //   2. Default severity Hidden/Info: We only execute analyzers reporting Warning/Error severity diagnostics on command line builds.
+            var analyzerShouldBeSkipped = !isEnabledByDefault ||
+                defaultSeverity == DiagnosticSeverity.Hidden ||
+                defaultSeverity == DiagnosticSeverity.Info;
+            var analyzer = new NamedTypeAnalyzerWithConfigurableEnabledByDefault(isEnabledByDefault, defaultSeverity, throwOnAllNamedTypes: analyzerShouldBeSkipped);
+            var diagnosticId = analyzer.Descriptor.Id;
+
+            var dir = Temp.CreateDirectory();
+            var src = dir.CreateFile("test.cs").WriteAllText(@"class C { }");
+
+            // Verify '/warnaserror-:DiagnosticId' behavior.
+            var args = new[] { "/warnaserror+", $"/warnaserror-:{diagnosticId}", "/nologo", "/t:library", "/preferreduilang:en", src.Path };
+
+            var cmd = CreateCSharpCompiler(null, dir.Path, args, analyzers: ImmutableArray.Create<DiagnosticAnalyzer>(analyzer));
+            var outWriter = new StringWriter(CultureInfo.InvariantCulture);
+            var exitCode = cmd.Run(outWriter);
+            var expectedExitCode = !analyzerShouldBeSkipped && defaultSeverity == DiagnosticSeverity.Error ? 1 : 0;
+            Assert.Equal(expectedExitCode, exitCode);
+
+            var output = outWriter.ToString();
+            if (analyzerShouldBeSkipped)
+            {
+                Assert.Empty(output);
+            }
+            else
+            {
+                var prefix = defaultSeverity == DiagnosticSeverity.Warning ? "warning" : "error";
+                Assert.Contains($"{prefix} {diagnosticId}: {analyzer.Descriptor.MessageFormat}", output);
+            }
+        }
+
         [Fact]
         public void SourceGenerators_EmbeddedSources()
         {

--- a/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs
+++ b/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs
@@ -321,7 +321,10 @@ namespace Microsoft.CodeAnalysis.Diagnostics
                 var isSuppressed = !diag.IsEnabledByDefault;
 
                 // Compilation wide user settings from ruleset/nowarn/warnaserror overrides the analyzer author.
-                if (diagnosticOptions.TryGetValue(diag.Id, out var severity))
+                // Note that "/warnaserror-:DiagnosticId" adds a diagnostic option with value 'ReportDiagnostic.Default',
+                // which should not alter 'isSuppressed'.
+                if (diagnosticOptions.TryGetValue(diag.Id, out var severity) &&
+                    severity != ReportDiagnostic.Default)
                 {
                     isSuppressed = severity == ReportDiagnostic.Suppress;
                 }

--- a/src/Compilers/VisualBasic/Test/CommandLine/CommandLineTests.vb
+++ b/src/Compilers/VisualBasic/Test/CommandLine/CommandLineTests.vb
@@ -10021,6 +10021,43 @@ End Class")
             End If
         End Sub
 
+        <WorkItem(47017, "https://github.com/dotnet/roslyn/issues/47017")>
+        <CombinatorialData, Theory>
+        Public Sub TestWarnAsErrorMinusDoesNotEnableDisabledByDefaultAnalyzers(defaultSeverity As DiagnosticSeverity, isEnabledByDefault As Boolean)
+            ' This test verifies that '/warnaserror-:DiagnosticId' does not affect if analyzers are executed or skipped.
+            ' Setup the analyzer to always throw an exception on analyzer callbacks for cases where we expect analyzer execution to be skipped:
+            '   1. Disabled by default analyzer, i.e. 'isEnabledByDefault == false'.
+            '   2. Default severity Hidden/Info: We only execute analyzers reporting Warning/Error severity diagnostics on command line builds.
+            Dim analyzerShouldBeSkipped = Not isEnabledByDefault OrElse
+                defaultSeverity = DiagnosticSeverity.Hidden OrElse
+                defaultSeverity = DiagnosticSeverity.Info
+
+            Dim analyzer = New NamedTypeAnalyzerWithConfigurableEnabledByDefault(isEnabledByDefault, defaultSeverity, throwOnAllNamedTypes:=analyzerShouldBeSkipped)
+            Dim diagnosticId = analyzer.Descriptor.Id
+
+            Dim dir = Temp.CreateDirectory()
+            Dim src = dir.CreateFile("test.cs").WriteAllText("
+Class C
+End Class")
+
+            ' Verify '/warnaserror-:DiagnosticId' behavior.
+            Dim args = {"/warnaserror+", $"/warnaserror-:{diagnosticId}", "/nologo", "/t:library", "/preferreduilang:en", src.Path}
+
+            Dim cmd = New MockVisualBasicCompiler(Nothing, dir.Path, args, analyzer)
+            Dim outWriter = New StringWriter(CultureInfo.InvariantCulture)
+            Dim exitCode = cmd.Run(outWriter)
+            Dim expectedExitCode = If(Not analyzerShouldBeSkipped AndAlso defaultSeverity = DiagnosticSeverity.[Error], 1, 0)
+            Assert.Equal(expectedExitCode, exitCode)
+
+            Dim output = outWriter.ToString()
+            If analyzerShouldBeSkipped Then
+                Assert.Empty(output)
+            Else
+                Dim prefix = If(defaultSeverity = DiagnosticSeverity.Warning, "warning", "error")
+                Assert.Contains($"{prefix} {diagnosticId}: {analyzer.Descriptor.MessageFormat}", output)
+            End If
+        End Sub
+
         <Theory, CombinatorialData>
         Public Sub TestAdditionalFileAnalyzer(registerFromInitialize As Boolean)
             Dim srcDirectory = Temp.CreateDirectory()

